src/ray/core_worker/common.h

```diff
@@ -20,6 +20,63 @@ struct RayFunction {
   const std::vector<std::string> function_descriptor;
 };
 
+/// Binary representation of ray object.
+class RayObject {
+ public:
+  /// Create a ray object instance.
+  ///
+  /// \param[in] data Data of the ray object.
+  /// \param[in] metadata Metadata of the ray object.
+  /// \param[in] copy_data Whether this class should hold a copy of data.
+  RayObject(const std::shared_ptr<Buffer> &data,
+            const std::shared_ptr<Buffer> &metadata = nullptr, bool copy_data = false)
+      : data_(data), metadata_(metadata), has_data_copy_(copy_data) {
+    if (has_data_copy_) {
```

Comment:
I know that this PR didn't introduce this flag, but can you please explain why we need the `has_data_copy_` flag and its semantics in the header comment?

Reply:
What happens when it's set, when should the caller set it, what could go wrong if it isn't set when it should be, etc.

Reply:
This is related to `Buffer` definition, please refer to src/ray/common/buffer.h.

For example, by default when you initialize a `LocalMemoryBuffer` with a data pointer and a length, it just assigns the pointer and length, but doesn't copy the data content, this is for performance reasons, but in this case the buffer cannot ensure data validity, it instead relies the lifetime passed in data pointer. 

This is fine for most cases - for example when you put an object into store, you create a temporary RayObject and don't want to do an extra copy. But in some cases you do want to always hold a valid data - for example, memory store uses `RayObject` to represent objects, in this case you actually want the object data to remain valid after user puts it into store.

Reply:
I see, that seems reasonable but we need to document it better (basically, the content of this comment). The comment in `src/ray/common/buffer.h` also doesn't have much more information.